packages/aws-cdk/lib/api/garbage-collector.ts

```diff
@@ -0,0 +1,484 @@
+  private bootstrapStackName: string;
+  private maxWaitTime: number;
+
+  public constructor(readonly props: GarbageCollectorProps) {
+    this.garbageCollectS3Assets = ['s3', 'all'].includes(props.type);
+    this.garbageCollectEcrAssets = ['ecr', 'all'].includes(props.type);
+
+    this.permissionToDelete = ['delete-tagged', 'full'].includes(props.action);
+    this.permissionToTag = ['tag', 'full'].includes(props.action);
+    this.maxWaitTime = props.maxWaitTime ?? 60000;
+
+    print(chalk.white(this.permissionToDelete, this.permissionToTag, props.action));
+
+    this.bootstrapStackName = props.bootstrapStackName ?? DEFAULT_TOOLKIT_STACK_NAME;
+
+    // TODO: ECR garbage collection
+    if (this.garbageCollectEcrAssets) {
+      throw new Error('ECR garbage collection is not yet supported');
+    }
+  }
+
+  /**
+   * Perform garbage collection on the resolved environment.
+   */
+  public async garbageCollect() {
+    print(chalk.black(this.garbageCollectS3Assets));
+    // SDKs
+    const sdk = (await this.props.sdkProvider.forEnvironment(this.props.resolvedEnvironment, Mode.ForWriting)).sdk;
+    const cfn = sdk.cloudFormation();
+    const s3 = sdk.s3();
+
+    const qualifier = await this.bootstrapQualifier(sdk, this.bootstrapStackName);
+
+    const activeAssets = new ActiveAssetCache();
+
+    let refreshStacksRunning = false;
+    const refreshStacks = async (isInitial?: boolean) => {
+      if (refreshStacksRunning) {
+        return;
+      }
+
+      refreshStacksRunning = true;
+
+      try {
+        const stacks = await this.fetchAllStackTemplates(cfn, qualifier);
+        for (const stack of stacks) {
+          activeAssets.rememberStack(stack);
+        }
+      } catch (err) {
+        throw new Error(`Error refreshing stacks: ${err}`);
+      } finally {
+        refreshStacksRunning = false;
+
+        if (!isInitial) {
+          setTimeout(refreshStacks, 300_000);
+        }
+      }
+    };
+
+    // Grab stack templates first
+    await refreshStacks(true);
+    // Refresh stacks in the background
+    const timeout = setTimeout(refreshStacks, 300_000);
+
+    try {
+      const bucket = await this.bootstrapBucketName(sdk, this.bootstrapStackName);
+      const numObjects = await this.numObjectsInBucket(s3, bucket);
+      const batches = 1;
+      const batchSize = 1000;
+      const currentTime = Date.now();
+      const graceDays = this.props.rollbackBufferDays;
+
+      print(chalk.white(`Parsing through ${numObjects} objects in batches`));
+
+      // Process objects in batches of 1000
+      // This is the batch limit of s3.DeleteObject and we intend to optimize for the "worst case" scenario
+      // where gc is run for the first time on a long-standing bucket where ~100% of objects are isolated.
+      for await (const batch of this.readBucketInBatches(s3, bucket, batchSize)) {
+        print(chalk.green(`Processing batch ${batches} of ${Math.floor(numObjects / batchSize) + 1}`));
```

Comment:
Should we pause here to give the background refresh an opportunity to stabilize?

```ts
await backgroundRefresh.waitNotRunning();
```

Reply:
wouldn't this open the door for a never ending loop if `backgroundRefresh` takes longer than the interval we set? If the interval is 5 mins and backgroundRefresh takes 6, then we never get passed a `waitNotRunnign`.